xbmc/pvr/PVRContextMenus.cpp

```diff
@@ -0,0 +1,730 @@
+      const CPVRRecordingPtr recording(item.GetPVRRecordingInfoTag());
+      if (recording && recording->IsDeleted())
+        return g_localizeStrings.Get(19291); /* Delete permanently */
+
+      return g_localizeStrings.Get(117); /* Delete */
+    }
+
+    bool DeleteRecording::IsVisible(const CFileItem &item) const
+    {
+      if (item.GetPVRRecordingInfoTag())
+        return true;
+
+      return false;
+    }
+
+    bool DeleteRecording::Execute(const CFileItemPtr &item) const
+    {
+      return CPVRGUIActions::GetInstance().DeleteRecording(item);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////
+    // Undelete recording
+
+    bool UndeleteRecording::IsVisible(const CFileItem &item) const
+    {
+      const CPVRRecordingPtr recording(item.GetPVRRecordingInfoTag());
+      if (recording && recording->IsDeleted())
+        return true;
+
+      return false;
+    }
+
+    bool UndeleteRecording::Execute(const CFileItemPtr &item) const
+    {
+      return CPVRGUIActions::GetInstance().UndeleteRecording(item);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////
+    // Activate / deactivate timer or timer rule
+
+    std::string ToggleTimerState::GetLabel(const CFileItem &item) const
+    {
+      const CPVRTimerInfoTagPtr timer(item.GetPVRTimerInfoTag());
+      if (timer && timer->m_state != PVR_TIMER_STATE_DISABLED)
+        return g_localizeStrings.Get(844); /* Deactivate */
+
+      return g_localizeStrings.Get(843); /* Activate */
+    }
+
+    bool ToggleTimerState::IsVisible(const CFileItem &item) const
+    {
+      const CPVRTimerInfoTagPtr timer(item.GetPVRTimerInfoTag());
+      if (!timer || URIUtils::PathEquals(item.GetPath(), CPVRTimersPath::PATH_ADDTIMER))
+        return false;
+
+      const CPVRTimerTypePtr timerType(timer->GetTimerType());
+      return timerType && timerType->SupportsEnableDisable();
+    }
+
+    bool ToggleTimerState::Execute(const CFileItemPtr &item) const
+    {
+      return CPVRGUIActions::GetInstance().ToggleTimerState(item);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////
+    // Add timer rule
+
+    bool AddTimerRule::IsVisible(const CFileItem &item) const
+    {
+      const CEpgInfoTagPtr epg(item.GetEPGInfoTag());
+      if (epg)
+        return g_PVRClients->SupportsTimers() && !epg->Timer();
+
+      return false;
+    }
+
+    bool AddTimerRule::Execute(const CFileItemPtr &item) const
+    {
+      return CPVRGUIActions::GetInstance().AddTimerRule(item, true);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////
+    // Edit timer rule
+
+    bool EditTimerRule::IsVisible(const CFileItem &item) const
+    {
+      CPVRTimerInfoTagPtr timer;
+
+      const CEpgInfoTagPtr epg(item.GetEPGInfoTag());
+      if (epg)
+        timer = epg->Timer();
+
+      if (!timer)
+        timer = item.GetPVRTimerInfoTag();
```

Comment:
Stupid github diff. The function above is not what you initially commented on, but the result of my refactoring...